lib/determinator/control.rb

```diff
@@ -115,12 +116,37 @@ def choose_target_group(feature, properties)
 
         tg.constraints.reduce(true) do |fit, (scope, *required)|
           present = [*normalised_properties[scope]]
-          fit && (required.flatten & present.flatten).any?
+          fit && (has_any(required, present) || has_any_app_version?(scope, required, present))
         end
-      # Must choose target group deterministically, if more than one match
+        # Must choose target group deterministically, if more than one match
       }.sort_by { |tg| tg.rollout }.last
     end
 
+    def has_any(required, present)
+      (required.flatten & present.flatten).any?
+    end
+
+    def has_any_app_version?(scope, required, present)
+      return false unless scope == 'app_version'
+
+      present.flatten.select {|version|
+        begin
+          present_sem_version = Semantic::Version.new version
+        rescue
+          # Ignore malformed app_version value
+          next false
+        end
+        required.flatten.all? {|constraint|
+          begin
+            present_sem_version.satisfies? constraint
+          rescue
+            # Ignore malformed constraint
+            next
```

Comment:
Same here?

Reply:
What's the exception here?

Reply:
```
     Failure/Error: Determinator.on_error { |e| error_logger.call(e) }
       #<Double "error_logger"> received unexpected message :call with (#<NoMethodError: undefined method `strip' for nil:NilClass>)
```
Should I catch `NoMethodError` instead?

Reply:
Now only rescuing `NoMethodError`

Reply:
You're getting a no method error because the previous rescue catches an exception when generating `present_sem_version`. You should probably code around that not being set, rather than catching an exception.

Reply:
I don't think so. I just tried this in a REPL:
```
begin
  present_sem_version = Semantic::Version.new "1.3.0"
  present_sem_version.satisfies? "whatever"
rescue NoMethodError
  p "Oh noes"
end
```
And it went through the rescue. I think `satisfies?` doesn't output a very clean error.

Reply:
It's not a very well written API from the `semantic` developers here — having a `?` method raising an exception (having `begin/rescue` blocks within a loop is a bit of a code smell to me) I'd actually suggest doing your validation steps first, then your test steps later (without any rescuing):

```ruby
begin
  present_sem_version = Semantic::Version.new "1.3.0"
  required_sem_version = Semantic::Version.new "1.4.0"
rescue NoMethodError
  p "Oh noes; your semvers ain't right"
end

# The looping bit

if present_sem_version.satisfies? required_sem_version.to_s
  puts "YATTA!"
end
```

(It's annoying that `satisfies?` doesn't accept a `Semantic::Version` 😑)

Reply:
Unfortunately I don’t think we can do this. `required` is a list of constraints and has the following format: `[">=1.2.3", "<=1.3.0"]`

So we can’t do `Semantic::Version.new constraint` and we need a loop :confused:

Reply:
Moved that bit of code in a private function for more clarity:
```ruby
    def satisfies?(version, constraint)
      version.satisfies? constraint
    rescue NoMethodError
      # Ignore malformed constraint
      false
    end
```

Reply:
Had a poke around the api - perhaps something like this?

```
  def has_any_app_version?(required, present)
      invalid_groups = present.flatten.select do |v|
        !v.match?(Semantic::Version::SemVerRegexp)
      end
      invalid_properties = required.flatten.select do |v|
        !v.match?(/\d/)
      end

      return false if (invalid_properties + invalid_groups).any?

      present.flatten.any? do |g|
        given_version = Semantic::Version.new(g)
        required.flatten.all? do |n|
          given_version.satisfies?(n)
        end
      end
    end
```

It's a little bit weird to pass in multiple versions - what would that mean in practise? Perhaps we should limit `present` to an array of one?

Reply:
The raise from `given_version.satisfies?(n)` if you pass it a string without a number in it just looks like a bug to me https://github.com/jlindsey/semantic/blob/master/lib/semantic/version.rb#L96

Reply:
> It's a little bit weird to pass in multiple versions - what would that mean in practise? Perhaps we should limit present to an array of one?

I guess there could be a rare case where we would want to have a flag enabled between one version and another one (e.g [">=1.2.3", "<=1.3.0"]). Although I agree it would be rare enough.

> The raise from given_version.satisfies?(n) if you pass it a string without a number in it just looks like a bug to me 

I'll `raise` an issue 👌 

> Had a poke around the api - perhaps something like this?

I'll be out the next few days but I'll give it a go when I'm back.

Reply:
@danielcooper I just found some time to test your code and it works fine 💯 
I find it cleaner so I'll go with that. Thanks for the suggestion!